### ISSUE TITLE
Add Storage Provider Durability metrics

### ISSUE THREAD

**Issue by davidgasquez** [davidgasquez/filecoin-data-portal#79]
We need a "durability" metric for Storage Providers. Right now, the metric is something like _Number of Successful PoSts_ divided by the _Expected Number of Successful PoSts_. 

 This number should be a trailing 30 days average.

There are several potential sources for this data.

- Lily is able to index [PoSt for all sectors](https://lilium.sh/data/actors/#miner_sector_posts), but no one is running this task (it’s a costly task to run and hard to digest useful information).
- Some of the PoSt data might be able via JSON-RPC methods. Getting historical (30 days) data might not be possible as [state is usually pruned](https://filecoinproject.slack.com/archives/CP50PPW2X/p1729148697654469?thread_ts=1728646177.881599&cid=CP50PPW2X), but this should be accurate as it comes directly from a node.
- There might be a way to compute successful PoSts by looking at parsed messages. E.g: parsing all  `SubmitWindowedPoST` messages that executed correctly. We’ll need access to a table with all messages indexed, and also make sure we haven't missed a message or a receipt.
- Lily’s [`miner_sector_events` could work to see `SECTOR_FAULTED`](https://lilium.sh/data/actors/#miner_sector_events) events per miner. This table has issues and is missing some of the events. Hard to say how off it is as there isn’t anything to compare against.

A bigger question here is: Knowing that Lily's data is not 100% accurate and requires heavy investment on infrastructure and a team to run, do we want to work with that data?

**Comment by BigLep** [comment]
@davidgasquez: circling back what about using [StateMinerFaults](https://lotus.filecoin.io/reference/lotus/state/#stateallminerfaults) or [StateAllMinerFaults](https://lotus.filecoin.io/reference/lotus/state/#stateallminerfaults) as [originally brought up in slack](https://filecoinproject.slack.com/archives/CP50PPW2X/p1729148570757399?thread_ts=1728646177.881599&cid=CP50PPW2X). I know this doesn't get us historical information, but I assume we could start doing daily probes here.  By default this would just tell us the amount of days an SP was in fault or not.  Ideally we'd have a way to scale this by the number of sectors that an SP had that day when probed.

In terms of getting historical information information, it [wasn't clear to me from Slack](https://filecoinproject.slack.com/archives/CP50PPW2X/p1729148697654469?thread_ts=1728646177.881599&cid=CP50PPW2X) whether an archival node will have this info.  It looks like [the API takes in the desired tipset](https://github.com/filecoin-project/lotus/blob/38e4d91409b0f81cf71cb4574d013e8a2809c00e/api/api_full.go#L405) to get the miner faults for.

**Comment by davidgasquez** [comment]
Great points @BigLep! We can't use [`StateAllMinerFaults`](https://lotus.filecoin.io/reference/lotus/state/#stateallminerfaults) as is not fully implemented and returns `fixme`. 

We could loop through all the SPs with power and get their [`StateMinerFaults`](https://lotus.filecoin.io/reference/lotus/state/#stateallminerfaults). This leaves us with two questions:

1. Do we need to get the data for all epochs / tipsets on a day? If not, we'd need to do 2880 calls per SP per day. That's around 6048000 calls with today's numbers.
2. Given 1 can be done once per day, it'll give us the number of faults. We still have to figure out how to compute the expected number of PoSts. I think this can be derived from the [miner partitions](https://github.com/filecoin-project/lotus/blob/master/api/v0api/full.go#L355), but still an open question.

**Comment by davidgasquez** [comment]
### Current State

We need a "durability" metric for Storage Providers. The metric we are exploring is:

$$\frac{\text{Number of Successful PoSts}}{\text{Expected Number of Successful PoSts}}$$

That can also be represented as:

$$\frac{\text{Expected Number of Successful PoSts - Number of Faults}}{\text{Expected Number of Successful PoSts}}$$

We've explored the following paths:

- Using Lily [`miner_sector_posts`](https://lilium.sh/data/actors/#miner_sector_posts). We encountered that no one is running this task as it's a costly task and hard to digest useful information from.
- Looking at the chain parsed messages. Will require getting and parsing all `SubmitWindowedPoST` messages that executed correctly. We’ll need access to a table with all messages indexed, and also make sure we haven't missed a message or a receipt.
- Using Lily [`miner_sector_events`](https://lilium.sh/data/actors/#miner_sector_events). We'll need all [SECTOR_FAULTED](https://lilium.sh/data/actors/#miner_sector_events) events per provider. This table has data quality issues and is missing some events.
- Using the JSON-RPC endpoints. 
  - We can get the faults by querying [StateMinerFaults](https://lotus.filecoin.io/reference/lotus/state/#stateallminerfaults) once per day per SP with power.
  - We'll need to figure out a way to get the _Expected Number of Successful PoSts_ from other methods (perhaps `MinerPartitions`?).


Right now, the most promising path is to rely on the JSON-RPC endpoints and run a dataset progressively. This will remove any dependency on Lily and make the entire pipeline very reproducible. We need to confirm these questions:

- It is possible to get a list of Storage Providers with active Sectors.
  - I have this list but is coming from Lily / Starboard. If we figure this out, we can make the entire dashboard rely on a JSON-RPC endpoint! 
- It is possible to get the expected number of PoSts for those Providers.
- It is enough to get the [StateMinerFaults](https://lotus.filecoin.io/reference/lotus/state/#stateallminerfaults) once per day per SP.

**Comment by davidgasquez** [comment]
After a great chat with Kuba (thanks for making that happen @BigLep), we have a better understanding of how the JSON-RPC approach and a potentially new definition of Durability.

Looking a PoSts might be too brittle and would also require parsing messages, as the `SubmitWindowedPoSt` method supports up to 3 partitions in the same message.

We should look at [`StateMinerPartitions`](https://lotus.filecoin.io/reference/lotus/state/#stateminerpartitions) for the Provider's active deadlines (obtained via [`StateMinerDeadlines`](https://lotus.filecoin.io/reference/lotus/state/#stateminerdeadlines)). If we do this twice per day per provider, we'll get a list of `FaultySectors`, `LiveSectors`, `ActiveSectors` sectors. Alternatively, we can derive the number of active sectors by getting the Provider Raw Power divided by their sector size.

The new formula to compute the Durability, would look like this over a period of time:

$$\frac{\text{Unique Active Sectors - Number of Faults}}{\text{Unique Active Sectors}}$$

The main challenge with this approach is that requires lots of call to the remote JSON-RPC that will take a while to complete.

Given this new simplified definition, it might be better to rely on Lily's [`miner_sector_events`](https://lilium.sh/data/actors/#miner_sector_events) table. We'll still need to figure out the number of Unique Active Sectors over a period of time, which seems to be the tricky part on both approaches. 

Focusing on that next!

**Comment by BigLep** [comment]
Thanks for recapping this @davidgasquez .  I think we should maybe structure the conversation like this:

### Ways to get the number of unique actors for a day:
1. Derive getting the Provider Raw Power divided by the provider sector size 
   * For each providerId, requires:
      * one call to `[StateMinerInfo](https://lotus.filecoin.io/reference/lotus/state/#stateminerinfo)`
      * one call to `[StateMinerPower](https://lotus.filecoin.io/reference/lotus/state/#stateminerpower)`
2. From looking across [StateMinerPartitions](https://lotus.filecoin.io/reference/lotus/state/#stateminerpartitions) for a given day's epoch:
   * For each providerId, requires:
       * one call to [StateMinerDeadlines](https://lotus.filecoin.io/reference/lotus/state/#stateminerdeadlines) to get active deadlines 
       * upwards of 48 calls to [StateMinerPartitions](https://lotus.filecoin.io/reference/lotus/state/#stateminerpartitions)
    * Note: this is the same call pattern that would be followed if `StateMinerPartitions` is used for getting the number of sectors that faulted for a day.  In this case though we're summing "AllSectors".

### Ways to get the number of sectors that faulted for a given day:
1. From looking across [StateMinerPartitions](https://lotus.filecoin.io/reference/lotus/state/#stateminerpartitions) for a given day's epoch
   * This would follow the same approach described above but in this case sum "FaultySectors"
2. Use Lily's [miner_sector_events](https://lilium.sh/data/actors/#miner_sector_events) table

By breaking it up this way, we can evaluate each option based on various dimensions like accuracy, cost to operate, time to execute, ease of backfilling, etc.

----
> We'll still need to figure out the number of Unique Active Sectors over a period of time

I agree this need to be figured out and is independent of faulty sectors.  

> which seems to be the tricky part on both approaches

"Derive getting the Provider Raw Power divided by the provider sector size" on the surface seems pretty straightforward given we're talking two RPC calls per SP per day.  

---
Also, can you capture where we landed about `[StateMinerInfo](https://lotus.filecoin.io/reference/lotus/state/#stateminerinfo)`?  I didn't take notes, the zoom recording didn't capture our breakout, and my head is foggy.  As a way to "get the number of sectors that faulted for a given day", can we just call `StateMinerInfo` once per day per SP (or are there followup calls that also need to be made)?

**Comment by BigLep** [comment]
Looking at https://colab.research.google.com/drive/1xUQmoHAPYh-DqJQMtmhRG3CUYl9NGIzq?usp=sharing for giving answer to questions listed there and raise other things coming to mind .  (Let me know if there is a better way to comment.)

> If we are going to look at the last 30 days, I think we need a distinct count of unique sectors. E.g: looking only twice per day might give us always the same number while in reality the provider might be terminating and adding new sectors.

I assume your concern is that we may not counting the actual number of active sectors.  For example, if over the course of day X a SP adds 1000 sectors and terminates 1000, there may have been a part day X where they actually had upwards of 2000 active sectors, but our metric may only see 1000 active sectors for day X.  I think this is a fair callout to have in our documentation, but I don't think it's important to optimize for given sectors aren't transient entities that SPs churn through quickly (repeatedly each day activating sectors and terminating them).  Over the course of a sector's lifecycel which I believe is usually on the order of months, this small time period where we're not counting fully accurately for the denominator isn't a big deal.

---
> all_active_sectors = set()
> all_fauly_sectors = set()

Sets are being used for looking at results across the previous 30 days.  I don't think that's the right data structure in that case.
I think we can use a set for a day's active and faulty sectors, but when looking across 30 days, we should be summing the size of each day's set.  That's probably not coming across well in English.  

Ideally we have:

Date | providerId | number of unique sectors that faulted | number of unique sectors
-- | -- | -- | --
2024-09-24 | f01713397 | 75192 | 86937
2024-09-25 | f01713397 | 9396 | 86937
2024-09-26 | f01713397 | 2349 | 86937
2024-09-27 | f01713397 | 0 | 86937
2024-09-28 | f01713397 | 0 | 86937

If we want to produce the durability looking at the last 5 days (doing 5 instead of 30 for simplicity), we do:

```
5*86937 - (75192+9396+2349)
---------------------------------
5*86937
```

Does that make sense?

Basically, I think the data tables you were generating from Lilly look like how I'd want to see us store the data.  We can then do trailing 30 day calculations on top of that data (assuming there isn't pushback on doing 30-day trailing calculations).  

---
> What happens if a sectors faults, then recover, and fault again? Neither looking at the set or the fault count will tell us that.

I think it depends on the duration that a sector is in faulty status and if we do a one-check-per-day approach.  Lets say we always sample for the first epoch of a given day in UTC.  Here is an example scenario:

date | epoch | what happens | note
-- | -- | -- | --
beginning of day X in UTC | Y | day X's metric calculation |  
part way through day X | Y + 100 | sector Z fault |  
beginning of day X + 1 in UTC | Y + 2880 | day X+1's metric calculation | detects sector Z fault
part way through day X + 1 | Y + 2980 | sector Z is healthy |  
beginning of day X + 2 in UTC | Y + 5760 | day X+2's metric calculation | see sector Z is healthy
part way through day X + 2 | Y + 5800 | sector Z fault |  
later on day X + 2 | Y + 5900 | sector Z is healthy |  
beginning of day X + 3 in UTC | Y + 8640 | day X+3's metric calculation | still sees sector Z is healthy.  It didn't capture the fault
near the end of day X + 3 | Y + 10640 | sector Z fault |  
beginning of day X + 4 in UTC | Y + 11250 | day X+4's metric calculation | detects sector Z fault
beginning of day X + 5 in UTC | Y + 14400 | day X+5's metric calculation | still detects sector Z fault
early on day X + 5 | Y + 15000 | sector Z is healthy |  
beginning of day X + 6 in UTC | Y + 17280 | day X+6's metric calculation | see sector Z is healthy


I think what you're calling out is that:
1. If sector fault duration lasts 24 hours our metric process is fine
2. If sector fault duration is less than 24 hours then we may miss it
3. If sector fault duration last more than 24 hours than we might overly penalize. 

I personally think item 3 is fine.  It's missing number 2 that I'm concerned about.

I think the key thing to track down here is what kind of durations a sector can can be faulted on.  Can they flip from faulted to healthy in less than 24 hours?  Basically the second question you raised "How much can a fault last in the faulty_sectors?"?

Basically I think the daily probe approach may have problems if faulty sector durations last less than 24 hours and we're using a "daily check at epoch 0 for that day" approach.  I think we're good though if we're able to look at `miner_sector_events` with lilly and then just sum up the events for a given day.  Basically your Lilly data looks really good to me.

Does that make sense?

**Comment by BigLep** [comment]
> The new formula to compute the Durability, would look like this over a period of time:

Oh, I think there's an important aspect to be clear about here.

I think this is accurate when looking for a single day:

```
Unique Active Sectors - Number of Faults
----------------------------------
Unique Active Sectors
```

When looking across days, I think the equation is 

```
Unique Active Sector Days - Number of Sector Fault Days
----------------------------------
Unique Active Sector Days
```

Does that make sense?

**Comment by davidgasquez** [comment]
Hey @BigLep! Thanks so much for getting back to me with such clarity and details. Appreciate how you keep track of everything and push back / share thoughts on things. 

As for comments, this is the place as it'll be useful for me in the future and open to anyone to join the discussion and refer back in a future if they need to know anything about Durability. :sweat_smile: 

Let me get back to you on all the things! 

---

> Also, can you capture where we landed about [`StateMinerInfo`](https://lotus.filecoin.io/reference/lotus/state/#stateminerinfo)?  I didn't take notes, the zoom recording didn't capture our breakout, and my head is foggy.  As a way to "get the number of sectors that faulted for a given day", can we just call `StateMinerInfo` once per day per SP (or are there followup calls that also need to be made)?

We can use `StateMinerInfo` to get `SectorSize` of a given provider, but I don't think we can get the number of sectors that faulted for a given day from this call.

> I assume your concern is that we may not counting the actual number of active sectors. For example, if over the course of day X a SP adds 1000 sectors and terminates 1000, there may have been a part day X where they actually had upwards of 2000 active sectors, but our metric may only see 1000 active sectors for day X. I think this is a fair callout to have in our documentation, but I don't think it's important to optimize for given sectors aren't transient entities that SPs churn through quickly (repeatedly each day activating sectors and terminating them). Over the course of a sector's lifecycel which I believe is usually on the order of months, this small time period where we're not counting fully accurately for the denominator isn't a big deal.

Thanks for the explanation! I was focusing on exploring all the possible places where we might be missing data and didn't explore / think as deeply about the actual impact of this missing data as you've done here.

> Basically, I think the data tables you were generating from Lilly look like how I'd want to see us store the data. We can then do trailing 30 day calculations on top of that data (assuming there isn't pushback on doing 30-day trailing calculations).

Oh, that makes sense! As you've seen, I've been using the number of unique active sectors in the last 30 days, not computing the daily values and then doing a trailing 30 day calculation. This approach then computes a "durability" metric at the daily level. We are adding the 30 days trailing calculation on top of that to smooth things out. Makes things simpler.

> I think the key thing to track down here is what kind of durations a sector can can be faulted on.  Can they flip from faulted to healthy in less than 24 hours?  Basically the second question you raised "How much can a fault last in the faulty_sectors?"?

Need to dig deeper into this. Will get back to you with an answer.

> Basically your Lilly data looks really good to me.

Yes! I think that's a promising approach. Yesterday I was exploring breath first to surface any issues quickly. Need to spend more time testing the data and checking the assumptions.

> I think this is accurate when looking for a single day:
> ```
> Unique Active Sectors - Number of Faults
> ----------------------------------
> Unique Active Sectors
> ```

On my mind, I was thinking about using this on the last 30 days window, but I see your point and I understand the change there. As I mentioned, it makes things simpler to think at the daily level and then add the 30 days trailing calculation on top of that.

---

Let me know if there is anything else you want me to clarify. Next up for me is:

- Audit `miner_sector_events`
- Benchmark the JSON-RPC method
- Get a few known `provider_id`s (not sure if anyone at FilOz runs a node), apply the metric and check how that feels with them.

**Comment by BigLep** [comment]
Thanks @davidgasquez!  I'll get a quick reply now and can circle back more later today (or maybe we'll get to connect sync).

> > Also, can you capture where we landed about [`StateMinerInfo`](https://lotus.filecoin.io/reference/lotus/state/#stateminerinfo)?  I didn't take notes, the zoom recording didn't capture our breakout, and my head is foggy.  As a way to "get the number of sectors that faulted for a given day", can we just call `StateMinerInfo` once per day per SP (or are there followup calls that also need to be made)?
> 
> We can use `StateMinerInfo` to get `SectorSize` of a given provider, but I don't think we can get the number of sectors that faulted for a given day from this call.

Doh - I knew I would make a typo like this. I mean to ask about [`StateFinerFaults`](https://lotus.filecoin.io/reference/lotus/state/#stateminerfaults).  I have updated the comment.

> * Get a few known `provider_id`s (not sure if anyone at FilOz runs a node), apply the metric and check how that feels with them.

Orjan and TippyFlits on slack from FilOz  both have provider ids I believe.

**Comment by davidgasquez** [comment]
Spent some time this morning exploring [`StateMinerFaults`](https://docs.filecoin.io/reference/json-rpc/state#stateminerfaults).

I'm still trying to understand what is going on, but I've found two things:

1. The data returned isn't either a bitfield or a list of unique sectors.

```bash
curl -X POST 'https://api.zondax.ch/fil/node/mainnet/rpc/v1' \
  -H 'Content-Type: application/json' \
  --data '{"jsonrpc":"2.0","id":1,"method":"Filecoin.StateMinerFaults","params": ["f01713397", [{"/": "bafy2bzacedkk552u4cp4nwvgdqxauylorrli5tangh4j6gwv6qylb6i3mqt6m"}, {"/": "bafy2bzacebta2cj3evr2rts3cv2e7syebtypv5jpzd7k27dp3vnvah4szat54"}, {"/": "bafy2bzacecbs3kuuxl7yfavlzvnqj7qktm7q4nr2gbpufw5at4gts26jzmrui"}, {"/": "bafy2bzacedidcaiefpbpirzfsbvjxsfsq5khyykwryuvokrvnntvn52fpexgk"}, {"/": "bafy2bzacec6aacli57ujqx7gugm624o4zdc4n3b5t6uvgznisth4sxlr6rvga"}, {"/": "bafy2bzacecxvddveznp5pyt2cpnrxjfaxorgmd5u3d4hjudvlvwwycmivyzx2"}]]}'
```

2. Doing one `StateMinerFaults` per hour per Provider doesn't return the same unique sectors failed in a day than the `StateMinerPartitions` (looping through deadlines) approach.

We'll need to figure out what is going on with these two issues and see if we can finally rely on `StateMinerFaults` to replace the 48 `StateMinerPartitions` we have to do for each miner.

**Comment by BigLep** [comment]
> We'll need to figure out what is going on with these two issues and see if we can finally rely on StateMinerFaults to replace the 48 StateMinerPartitions we have to do for each miner.

@davidgasquez and I agreed this isn't blocking as he can use `StateMinerPartitions` approach to validate [`miner_sector_events`](https://lilium.sh/data/actors/#miner_sector_events).

It would be nice to come to conclusion on whether we can use [`StateMinerFaults`](https://docs.filecoin.io/reference/json-rpc/state#stateminerfaults) because:
1. Validation will be faster given can make ~1 RPC call per day per SP rather than ~48 calls.
2. Close this open thread so it's understood for the future.

Example notebook comparing the two approaches: https://colab.research.google.com/drive/16yfD7G2aPzuEXUBwjCjMdBpkDbQOaDli#scrollTo=rLgG5Xu8TMig
#fil-lotus-help thread asking about this: https://filecoinproject.slack.com/archives/CPFTWMY7N/p1729887425504539

**Comment by BigLep** [comment]
Concerning 
> what kind of durations a sector can can be faulted on. Can they flip from faulted to healthy in less than 24 hours? Basically the second question you raised "How much can a fault last in the faulty_sectors?"?

From #fil-protocol: https://filecoinproject.slack.com/archives/C04USV0L7PF/p1729792940600989
> partitions can only be proven in their respective deadline window. so its 24h'ish

**Comment by beck-8** [comment]
> Spent some time this morning exploring [`StateMinerFaults`](https://docs.filecoin.io/reference/json-rpc/state#stateminerfaults).今天早上花了一些时间探索 [`StateMinerFaults`](https://docs.filecoin.io/reference/json-rpc/state#stateminerfaults)。
> 
> I'm still trying to understand what is going on, but I've found two things:我仍在努力了解发生了什么，但我发现了两件事：
> 
> 1. The data returned isn't either a bitfield or a list of unique sectors.返回的数据既不是位域，也不是唯一扇区的列表。
> 
> ```shell
> curl -X POST 'https://api.zondax.ch/fil/node/mainnet/rpc/v1' \
>   -H 'Content-Type: application/json' \
>   --data '{"jsonrpc":"2.0","id":1,"method":"Filecoin.StateMinerFaults","params": ["f01713397", [{"/": "bafy2bzacedkk552u4cp4nwvgdqxauylorrli5tangh4j6gwv6qylb6i3mqt6m"}, {"/": "bafy2bzacebta2cj3evr2rts3cv2e7syebtypv5jpzd7k27dp3vnvah4szat54"}, {"/": "bafy2bzacecbs3kuuxl7yfavlzvnqj7qktm7q4nr2gbpufw5at4gts26jzmrui"}, {"/": "bafy2bzacedidcaiefpbpirzfsbvjxsfsq5khyykwryuvokrvnntvn52fpexgk"}, {"/": "bafy2bzacec6aacli57ujqx7gugm624o4zdc4n3b5t6uvgznisth4sxlr6rvga"}, {"/": "bafy2bzacecxvddveznp5pyt2cpnrxjfaxorgmd5u3d4hjudvlvwwycmivyzx2"}]]}'
> ```
> 
> 2. Doing one `StateMinerFaults` per hour per Provider doesn't return the same unique sectors failed in a day than the `StateMinerPartitions` (looping through deadlines) approach.与 `StateMinerPartitions`（循环遍历截止时间）方法相比，每个 Provider 每小时执行一次 `StateMinerFaults` 不会返回一天内失败的相同唯一扇区。
> 
> We'll need to figure out what is going on with these two issues and see if we can finally rely on `StateMinerFaults` to replace the 48 `StateMinerPartitions` we have to do for each miner.我们需要弄清楚这两个问题是怎么回事，看看我们最终是否可以依靠 `StateMinerFaults` 来替换我们必须为每个矿工做的 48 个 `StateMinerPartitions`。

> 1. The data returned isn't either a bitfield or a list of unique sectors

The result is a bitfield

**Comment by beck-8** [comment]
The simplest approach is to iterate over the 48 deadline `StateMinerPartitions` and then aggregate the data.
`LiveSectors/(LiveSectors+FaultySectors+RecoveringSectors)`

I think this is the result you want. You can run it once, twice, or four times a day, depending on the data granularity you want.

**Comment by BigLep** [comment]
Thanks @beck-8.  I found that the library we were using wasn't handling the Run Length Encoding of BitFields when making a `StateMinerFaults` call.  Once I add that in, we're good.

@davidgasquez : I did a notebook update in https://colab.research.google.com/drive/16yfD7G2aPzuEXUBwjCjMdBpkDbQOaDli#scrollTo=poUgJ7Hcedxj&line=4&uniqifier=1 .  Once I handle the BitField properly the results look right.  As a result, I think you'll be good to switch to that approach rather than the `StateMinerPartitions` approach.
